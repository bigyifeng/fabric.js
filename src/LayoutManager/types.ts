import type {
  BasicTransformEvent,
  ModifiedEvent,
  TModificationEvents,
} from '../EventTypeDefs';
import type { Point } from '../Point';
import type { Group } from '../shapes/Group';
import type { ITextEvents } from '../shapes/IText/ITextBehavior';
import type { FabricObject } from '../shapes/Object/FabricObject';
import type { LayoutStrategy } from './LayoutStrategies/LayoutStrategy';

export type LayoutTrigger =
  | 'initialization'
  | 'object_modifying'
  | 'object_modified'
  | 'added'
  | 'removed'
  | 'imperative';

/**
 * positioning and layout data **relative** to instance's parent
 */
export type LayoutStrategyResult = {
  /**
   * new centerX as measured by the **containing** plane (same as `left` with `originX` set to `center`)
   */
  centerX: number;
  /**
   * new centerY as measured by the **containing** plane (same as `top` with `originY` set to `center`)
   */
  centerY: number;
  /**
   * correctionX to translate objects by, measured as `centerX`
   *
   * Since objects are measured relative to the group's center, once the group's size changes we must apply a correction to
   * the objects' positions so that they relate to the new center.
   * In other words, this value makes it possible to layout objects relative to the tl corner, for instance, but not only.
   *
   * @see `relativeCorrectionX`
   */
  correctionX?: number;
  /**
   * correctionY to translate objects by, measured as `centerY`
   *
   * Since objects are measured relative to the group's center, once the group's size changes we must apply a correction to
   * the objects' positions so that they relate to the new center.
   * In other words, this value makes it possible to layout objects relative to the tl corner, for instance, but not only.
   *
   * @see `relativeCorrectionY`
   */
  correctionY?: number;
  /**
   * correctionX to translate objects by as measured by the plane
   * @see `correctionX`
   */
  relativeCorrectionX?: number;
  /**
   * correctionY to translate objects by as measured by the plane
   * @see `correctionY`
   */
  relativeCorrectionY?: number;
  /**
   * new width of instance
   */
  width: number;
  /**
   * new height of instance
   */
  height: number;
};

export type LayoutResult = {
  result?: LayoutStrategyResult;
  prevCenter: Point;
  nextCenter: Point;
  /**
   * The vector used to offset objects by, as measured by the plane
   */
  offset: Point;
};

export type ImperativeLayoutOptions = {
  strategy?: LayoutStrategy;
  overrides?: LayoutStrategyResult;
};

export type CommonLayoutContext = {
  target: Group;
  strategy?: LayoutStrategy;
  type: LayoutTrigger;
  /**
   * array of objects starting from the object that triggered the call to the current one
   */
  path?: Group[];
};

export type InitializationLayoutContext = CommonLayoutContext & {
  type: 'initialization';
  objectsRelativeToGroup?: boolean;
  targets: FabricObject[];
  x?: number;
  y?: number;
};

export type CollectionChangeLayoutContext = CommonLayoutContext & {
  type: 'added' | 'removed';
  targets: FabricObject[];
};

export type ObjectModifiedLayoutContext = CommonLayoutContext & {
  type: 'object_modified';
  trigger: 'modified';
  e: ModifiedEvent;
};

export type ObjectModifyingLayoutContext =
  | CommonLayoutContext & {
      type: 'object_modifying';
    } & (
        | {
            trigger: TModificationEvents;
            e: BasicTransformEvent & { target: FabricObject };
          }
        | {
            trigger: 'text:changed';
            e: ITextEvents['changed'] & { target: FabricObject };
          }
      );

export type ImperativeLayoutContext = CommonLayoutContext & {
  type: 'imperative';
  overrides?: LayoutStrategyResult;
};

export type LayoutContext =
  | InitializationLayoutContext
  | CollectionChangeLayoutContext
  | ObjectModifiedLayoutContext
  | ObjectModifyingLayoutContext
  | ImperativeLayoutContext;

export type StrictLayoutContext = LayoutContext & {
  strategy: LayoutStrategy;
  prevStrategy?: LayoutStrategy;
  strategyChange: boolean;
};

export type LayoutBeforeEvent = {
  context: StrictLayoutContext;
};

export type LayoutEvent = {
  context: StrictLayoutContext;
  /**
   * will be undefined if layout was skipped
   */
  result?: LayoutResult;
};
