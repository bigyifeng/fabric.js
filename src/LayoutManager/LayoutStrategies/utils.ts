import { Point, ZERO } from '../../Point';
import type { Group } from '../../shapes/Group';
import type { FabricObject } from '../../shapes/Object/FabricObject';
import { multiplyTransformMatrixArray } from '../../util/misc/matrix';
import { sizeAfterTransform } from '../../util/misc/objectTransforms';
import {
  calcPlaneChangeMatrix,
  sendVectorToPlane,
} from '../../util/misc/planeChange';

/**
 * @returns 2 points, the tl and br corners of the non rotated bounding box of an object
 * in the {@link group} plane, taking into account objects that {@link group} is their parent
 * but also belong to the active selection.
 */
export const getObjectBounds = (
  group: Group,
  object: FabricObject
): Point[] => {
  const t =
    object.group && object.group !== group
      ? calcPlaneChangeMatrix(
          object.group.calcTransformMatrix(),
          group.calcTransformMatrix()
        )
      : null;
  const objectCenter = t
    ? object.getRelativeCenterPoint().transform(t)
    : object.getRelativeCenterPoint();
  const strokeUniformVector = object.strokeUniform
    ? sendVectorToPlane(
        new Point(object.strokeWidth, object.strokeWidth),
        undefined,
        group.calcTransformMatrix()
      )
    : ZERO;
  const scalingStrokeWidth = !object.strokeUniform ? object.strokeWidth : 0;
  const sizeVector = sizeAfterTransform(
    object.width + scalingStrokeWidth,
    object.height + scalingStrokeWidth,
    multiplyTransformMatrixArray([t, object.calcOwnMatrix()], true)
  )
    .add(strokeUniformVector)
    .scalarDivide(2);
  return [objectCenter.subtract(sizeVector), objectCenter.add(sizeVector)];
};
