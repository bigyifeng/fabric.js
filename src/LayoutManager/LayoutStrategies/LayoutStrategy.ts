import { Point } from '../../Point';
import type { FabricObject } from '../../shapes/Object/FabricObject';
import { makeBoundingBoxFromPoints } from '../../util/misc/boundingBoxFromPoints';
import { resolveOrigin } from '../../util/misc/resolveOrigin';
import type { LayoutStrategyResult, StrictLayoutContext } from '../types';
import { getObjectBounds } from './utils';

export abstract class LayoutStrategy {
  /**
   * override by subclass for persistence (TS does not support `static abstract`)
   */
  static type = 'strategy';

  shouldPerformLayout(context: StrictLayoutContext) {
    return (
      context.type === 'initialization' ||
      context.type === 'imperative' ||
      (!!context.prevStrategy && context.strategy !== context.prevStrategy)
    );
  }

  shouldLayoutClipPath({ type, target: { clipPath } }: StrictLayoutContext) {
    return (
      type !== 'initialization' && clipPath && !clipPath.absolutePositioned
    );
  }

  /**
   * called from the `onAfterLayout` hook
   */
  shouldResetTransform(context: StrictLayoutContext) {
    return context.target.size() === 0;
  }

  calcLayoutResult(
    context: StrictLayoutContext,
    objects: FabricObject[]
  ): LayoutStrategyResult | undefined {
    if (this.shouldPerformLayout(context)) {
      return this.calcBoundingBox(objects, context);
    }
  }

  /**
   * Override this method to customize layout.
   */
  calcBoundingBox(
    objects: FabricObject[],
    context: StrictLayoutContext
  ): LayoutStrategyResult | undefined {
    if (context.type === 'imperative' && context.overrides) {
      return context.overrides;
    }
    if (objects.length === 0) {
      return;
    }
    const { target } = context;
    const { left, top, width, height } = makeBoundingBoxFromPoints(
      objects.map((object) => getObjectBounds(target, object)).flat()
    );
    const size = new Point(width, height);
    const bboxCenter = new Point(left, top).add(size.scalarDivide(2));
    if (context.type === 'initialization') {
      // translate the layout origin from left top to target's origin
      const origin = bboxCenter.add(size.scalarMultiply(-0.5));
      const center = origin.add(
        size.multiply(
          new Point(
            -resolveOrigin(target.originX),
            -resolveOrigin(target.originY)
          )
        )
      );
      return {
        // in `initialization` we do not account for target's transformation matrix
        center,
        relativeCorrection: center.subtract(bboxCenter),
        size,
      };
    } else {
      //  we send `relativeCenter` up to group's containing plane
      const center = bboxCenter.transform(target.calcOwnMatrix());
      return {
        center,
        size,
      };
    }
  }
}
