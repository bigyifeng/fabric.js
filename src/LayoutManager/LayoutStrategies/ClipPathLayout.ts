import type { FabricObject } from '../../shapes/Object/FabricObject';
import { sendPointToPlane } from '../../util/misc/planeChange';
import type { LayoutStrategyResult, StrictLayoutContext } from '../types';
import { LayoutStrategy } from './LayoutStrategy';

export class ClipPathLayout extends LayoutStrategy {
  shouldPerformLayout(context: StrictLayoutContext): boolean {
    return (
      !!context.target.clipPath &&
      (context.type === 'initialization' || super.shouldPerformLayout(context))
    );
  }

  shouldLayoutClipPath() {
    return false;
  }

  calcLayoutResult(
    context: StrictLayoutContext,
    objects: FabricObject[]
  ): LayoutStrategyResult | undefined {
    const { target } = context;
    const { clipPath } = target;
    if (!clipPath || !this.shouldPerformLayout(context)) {
      return;
    }
    const bbox = this.calcBoundingBox([clipPath], context)!;
    if (clipPath.absolutePositioned) {
      //  we want the center point to exist in group's containing plane
      const clipPathCenter = sendPointToPlane(
        clipPath.getRelativeCenterPoint(),
        undefined,
        target.group?.calcTransformMatrix()
      );
      return {
        centerX: clipPathCenter.x,
        centerY: clipPathCenter.y,
        width: bbox.width,
        height: bbox.height,
      };
    } else {
      //  we want the center point to exist in group's containing plane, so we send it upwards
      const clipPathCenter = clipPath
        .getRelativeCenterPoint()
        .transform(target.calcOwnMatrix(), true);
      if (this.shouldPerformLayout(context)) {
        // the clip path is positioned relative to the group's center which is affected by the bbox
        // so we first calculate the bbox
        const {
          centerX = 0,
          centerY = 0,
          correctionX = 0,
          correctionY = 0,
        } = this.calcBoundingBox(objects, context) || {};
        return {
          centerX: centerX + clipPathCenter.x,
          centerY: centerY + clipPathCenter.y,
          correctionX: correctionX - clipPathCenter.x,
          correctionY: correctionY - clipPathCenter.y,
          width: clipPath.width,
          height: clipPath.height,
        };
      } else {
        const center = target.getRelativeCenterPoint();
        return {
          centerX: center.x + clipPathCenter.x,
          centerY: center.y + clipPathCenter.y,
          width: bbox.width,
          height: bbox.height,
        };
      }
    }
  }
}
