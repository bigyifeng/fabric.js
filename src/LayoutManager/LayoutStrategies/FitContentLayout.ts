import type { StrictLayoutContext } from '../types';
import { LayoutStrategy } from './LayoutStrategy';

export class FitContentLayout extends LayoutStrategy {
  static readonly type = 'fit-content';

  /**
   * @override layout on all triggers
   * Override at will
   */
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  shouldPerformLayout(context: StrictLayoutContext) {
    return true;
  }
}
