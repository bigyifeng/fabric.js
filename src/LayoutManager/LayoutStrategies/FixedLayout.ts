import type { FabricObject } from '../../shapes/Object/FabricObject';
import type { LayoutStrategyResult, StrictLayoutContext } from '../types';
import { LayoutStrategy } from './LayoutStrategy';

export class FixedLayout extends LayoutStrategy {
  calcBoundingBox(
    objects: FabricObject[],
    context: StrictLayoutContext
  ): LayoutStrategyResult | undefined {
    const result = super.calcBoundingBox(objects, context);
    if (context.type === 'initialization' && result) {
      const {
        target: { width, height },
      } = context;
      return {
        ...result,
        width: width || result.width,
        height: height || result.height,
      };
    }
    return result;
  }
}
