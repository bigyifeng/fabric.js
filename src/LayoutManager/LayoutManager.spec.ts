import { Point } from '../Point';
import { Group } from '../shapes/Group';
import { FabricObject } from '../shapes/Object/FabricObject';
import { LayoutManager } from './LayoutManager';
import { FitContentLayout } from './LayoutStrategies/FitContentLayout';
import type {
  LayoutContext,
  LayoutTrigger,
  StrictLayoutContext,
} from './types';

describe('Layout Manager', () => {
  it('should set fit content strategy by default', () => {
    expect(new LayoutManager().strategy).toBeInstanceOf(FitContentLayout);
  });

  describe('Lifecycle', () => {
    test.each([true, false])('performLayout with result of %s', (result) => {
      const lifecycle: jest.SpyInstance[] = [];
      const layoutResult = result
        ? {
            result: { centerX: 0, centerY: 0, width: 0, height: 0 },
            prevCenter: new Point(),
            nextCenter: new Point(),
            offset: new Point(),
          }
        : undefined;

      const manager = new LayoutManager();
      const onBeforeLayout = jest
        .spyOn(manager, 'onBeforeLayout')
        .mockImplementation(() => {
          lifecycle.push(onBeforeLayout);
        });
      const getLayoutResult = jest
        .spyOn(manager, 'getLayoutResult')
        .mockImplementation(() => {
          lifecycle.push(getLayoutResult);
          return layoutResult;
        });
      const commitLayout = jest
        .spyOn(manager, 'commitLayout')
        .mockImplementation(() => {
          lifecycle.push(commitLayout);
        });
      const onAfterLayout = jest
        .spyOn(manager, 'onAfterLayout')
        .mockImplementation(() => {
          lifecycle.push(onAfterLayout);
        });

      const targets = [new Group([new FabricObject()]), new FabricObject()];
      const target = new Group(targets);
      manager.performLayout({
        type: 'initialization',
        target,
        targets,
      });

      const context = {
        bubbles: true,
        strategy: manager.strategy,
        type: 'initialization',
        target,
        targets,
        prevStrategy: undefined,
      };

      const expectedLifecycle = [
        onBeforeLayout,
        getLayoutResult,
        ...(result ? [commitLayout] : []),
        onAfterLayout,
      ];

      expect(lifecycle).toEqual(expectedLifecycle);
      expectedLifecycle.forEach(
        ({
          mock: {
            calls: [[arg0]],
          },
        }) => expect(arg0).toMatchObject(context)
      );
      if (result) {
        [commitLayout, onAfterLayout].forEach(
          ({
            mock: {
              calls: [[arg0, arg1]],
            },
          }) => expect(arg1).toEqual(layoutResult)
        );
      } else {
        [onAfterLayout].forEach(
          ({
            mock: {
              calls: [[arg0, arg1]],
            },
          }) => expect(arg1).toBeUndefined()
        );
      }
    });

    it('should guard initial layout', () => {
      const manager = new LayoutManager();
      const onBeforeLayout = jest.spyOn(manager, 'onBeforeLayout');
      const target = new Group();
      manager.performLayout({ type: 'imperative', target });
      expect(onBeforeLayout).not.toHaveBeenCalled();
      manager.performLayout({
        type: 'initialization',
        target,
        targets: [],
      });
      expect(onBeforeLayout).toHaveBeenCalled();
    });
  });

  describe('onBeforeLayout', () => {
    describe('triggers', () => {
      const triggers = [
        'modified',
        'moving',
        'resizing',
        'rotating',
        'scaling',
        'skewing',
        'changed',
      ] as const;

      it('should subscribe object', () => {
        const lifecycle: jest.SpyInstance[] = [];
        const manager = new LayoutManager();
        const unsubscribe = jest
          .spyOn(manager, 'unsubscribe')
          .mockImplementation(() => {
            lifecycle.push(unsubscribe);
          });
        const object = new FabricObject();
        const on = jest.spyOn(object, 'on').mockImplementation(() => {
          lifecycle.push(on);
        });

        manager.subscribe({}, object);

        expect(lifecycle).toEqual([
          unsubscribe,
          ...new Array(triggers.length).fill(on),
        ]);
        expect(on.mock.calls.map(([arg0]) => arg0)).toEqual(triggers);
      });

      it('a subscribed object should trigger layout', () => {
        const manager = new LayoutManager();
        const performLayout = jest.spyOn(manager, 'performLayout');
        const object = new FabricObject();
        const target = new Group();
        manager.subscribe({ target }, object);

        const event = { foo: 'bar' };
        triggers.forEach((trigger) => object.fire(trigger, event));
        expect(performLayout.mock.calls).toMatchObject([
          [
            {
              e: { target: object, ...event },
              target,
              trigger: 'modified',
              type: 'object_modified',
            },
          ],
          ...triggers.slice(1).map((trigger) => [
            {
              e: { target: object, ...event },
              target,
              trigger,
              type: 'object_modifying',
            },
          ]),
        ]);

        performLayout.mockClear();
        expect(manager._subscriptions.get(object)).toBeDefined();
        manager.unsubscribe({ target }, object);
        expect(manager._subscriptions.get(object)).toBeUndefined();
        triggers.forEach((trigger) => object.fire(trigger, event));
        expect(performLayout).not.toHaveBeenCalled();
      });
    });

    it.each([
      { trigger: 'initialization', action: 'subscribe' },
      { trigger: 'added', action: 'subscribe' },
      { trigger: 'removed', action: 'unsubscribe' },
    ] as {
      trigger: LayoutTrigger;
      action: 'subscribe' | 'unsubscribe';
    }[])(
      '$trigger trigger should $action targets and call target hooks',
      ({ action }) => {
        const lifecycle: jest.SpyInstance[] = [];
        const targets = [new Group([new FabricObject()]), new FabricObject()];
        const target = new Group(targets);
        const targetOnBeforeLayout = jest
          .spyOn(target, 'onBeforeLayout')
          .mockImplementation(() => {
            lifecycle.push(targetOnBeforeLayout);
          });
        const targetFire = jest.spyOn(target, 'fire').mockImplementation(() => {
          lifecycle.push(targetFire);
        });

        const manager = new LayoutManager();
        const subscription = jest
          .spyOn(manager, action)
          .mockImplementation(() => {
            lifecycle.push(subscription);
          });

        const context: StrictLayoutContext = {
          bubbles: true,
          strategy: manager.strategy,
          type: 'initialization',
          target,
          targets,
          prevStrategy: undefined,
          stopPropagation() {
            this.bubbles = false;
          },
        };
        manager.onBeforeLayout(context);

        expect(lifecycle).toEqual([
          subscription,
          subscription,
          targetOnBeforeLayout,
          targetFire,
        ]);
        expect(targetOnBeforeLayout).toBeCalledWith({ context });
        expect(targetFire).toBeCalledWith('layout:before', {
          context,
        });
      }
    );
  });

  describe('getLayoutResult', () => {
    test.each([
      { type: 'initialization', targets: [] },
      { type: 'initialization', objectsRelativeToGroup: true, targets: [] },
      { type: 'imperative' },
    ] as const)('#type trigger', (options) => {
      const manager = new LayoutManager();
      jest.spyOn(manager.strategy, 'calcLayoutResult').mockReturnValue({
        centerX: 50,
        centerY: 100,
        width: 200,
        height: 250,
        correctionX: 10,
        correctionY: 20,
        relativeCorrectionX: -30,
        relativeCorrectionY: -40,
      });

      const target = new Group([], { scaleX: 2, scaleY: 0.5, angle: 30 });

      const context: StrictLayoutContext = {
        bubbles: true,
        strategy: manager.strategy,
        target,
        ...options,
        stopPropagation() {
          this.bubbles = false;
        },
      };

      expect(manager.getLayoutResult(context)).toMatchSnapshot();
    });
  });

  describe('commitLayout', () => {
    const prepareTest = (
      contextOptions: {
        type: 'initialization' | 'added';
      } & Partial<LayoutContext>
    ) => {
      const lifecycle: jest.SpyInstance[] = [];

      const targets = [new Group([new FabricObject()]), new FabricObject()];
      const target = new Group(targets, { strokeWidth: 0 });
      const targetSet = jest.spyOn(target, 'set').mockImplementation(() => {
        lifecycle.push(targetSet);
      });
      const targetSetCoords = jest
        .spyOn(target, 'setCoords')
        .mockImplementation(() => {
          lifecycle.push(targetSetCoords);
        });
      const targetSetPositionByOrigin = jest
        .spyOn(target, 'setPositionByOrigin')
        .mockImplementation(() => {
          lifecycle.push(targetSetPositionByOrigin);
        });

      const manager = new LayoutManager();
      const layoutObjects = jest
        .spyOn(manager, 'layoutObjects')
        .mockImplementation(() => {
          lifecycle.push(layoutObjects);
        });

      const context: StrictLayoutContext = {
        ...contextOptions,
        bubbles: true,
        strategy: manager.strategy,
        target,
        targets,
        prevStrategy: undefined,
        stopPropagation() {
          this.bubbles = false;
        },
      };

      return {
        manager,
        context,
        layoutResult: {
          result: { centerX: 5, centerY: 5, width: 10, height: 10 },
          prevCenter: new Point(),
          nextCenter: new Point(5, 5),
          offset: new Point(-5, -5),
        },
        targetMocks: {
          set: targetSet,
          setCoords: targetSetCoords,
          setPositionByOrigin: targetSetPositionByOrigin,
        },
        mocks: {
          layoutObjects,
        },
        lifecycle,
      };
    };

    it.each([{}, { x: 10 }, { y: 10 }, { x: 10, y: 10 }] as const)(
      'initialization trigger with %s should set size and position',
      (pos) => {
        const {
          manager,
          context,
          layoutResult,
          mocks: { layoutObjects },
          targetMocks,
          lifecycle,
        } = prepareTest({ type: 'initialization', ...pos });
        const {
          result: { width, height },
        } = layoutResult;

        manager.commitLayout(context, layoutResult);

        expect(lifecycle).toEqual([
          targetMocks.set,
          layoutObjects,
          targetMocks.set,
        ]);
        expect(targetMocks.set).nthCalledWith(1, { width, height });
        expect(layoutObjects).toBeCalledWith(context, layoutResult);
        expect(targetMocks.set).nthCalledWith(2, {
          left: pos.x ?? 0,
          top: pos.y ?? 0,
        });
      }
    );

    it('non initialization trigger should set size, position and invalidate target', () => {
      const {
        manager,
        context,
        layoutResult,
        mocks: { layoutObjects },
        targetMocks,
        lifecycle,
      } = prepareTest({ type: 'added' });
      const {
        result: { width, height },
      } = layoutResult;

      manager.commitLayout(context, layoutResult);

      expect(lifecycle).toEqual([
        targetMocks.set,
        layoutObjects,
        targetMocks.setPositionByOrigin,
        targetMocks.setCoords,
        targetMocks.set,
      ]);
      expect(targetMocks.set).nthCalledWith(1, { width, height });
      expect(layoutObjects).toBeCalledWith(context, layoutResult);
      expect(targetMocks.set).nthCalledWith(2, {
        dirty: true,
      });
    });
  });

  describe('onAfterLayout', () => {
    it.each([true, false])(
      'should call target hooks with bubbling %s',
      (bubbles) => {
        const lifecycle: jest.SpyInstance[] = [];
        const targets = [new Group([new FabricObject()]), new FabricObject()];
        const target = new Group(targets);
        const targetOnAfterLayout = jest
          .spyOn(target, 'onAfterLayout')
          .mockImplementation(() => {
            lifecycle.push(targetOnAfterLayout);
          });
        const targetFire = jest.spyOn(target, 'fire').mockImplementation(() => {
          lifecycle.push(targetFire);
        });

        const parent = new Group([target]);
        const parentPerformLayout = jest
          .spyOn(parent.layoutManager, 'performLayout')
          .mockImplementation(() => {
            lifecycle.push(parentPerformLayout);
          });

        const manager = new LayoutManager();
        const shouldResetTransform = jest
          .spyOn(manager.strategy, 'shouldResetTransform')
          .mockImplementation(() => {
            lifecycle.push(shouldResetTransform);
          });

        const context: StrictLayoutContext = {
          bubbles,
          strategy: manager.strategy,
          type: 'added',
          target,
          targets,
          prevStrategy: undefined,
          stopPropagation() {
            this.bubbles = false;
          },
        };
        const layoutResult = {
          result: { centerX: 0, centerY: 0, width: 0, height: 0 },
          prevCenter: new Point(),
          nextCenter: new Point(),
          offset: new Point(),
        };
        manager.onAfterLayout(context, layoutResult);

        expect(lifecycle).toEqual([
          shouldResetTransform,
          targetOnAfterLayout,
          targetFire,
          ...(bubbles ? [parentPerformLayout] : []),
        ]);
        expect(shouldResetTransform).toBeCalledWith(context);
        expect(targetOnAfterLayout).toBeCalledWith({
          context,
          result: layoutResult,
        });
        expect(targetFire).toBeCalledWith('layout', {
          context,
          result: layoutResult,
        });
        bubbles &&
          expect(parentPerformLayout.mock.calls[0]).toMatchObject([
            {
              type: 'added',
              targets,
              target: parent,
              path: [target],
            },
          ]);
      }
    );

    test.each([true, false])('reset target transform %s', (reset) => {
      const targets = [new Group([new FabricObject()]), new FabricObject()];
      const target = new Group(targets);
      target.left = 50;

      const manager = new LayoutManager();
      jest
        .spyOn(manager.strategy, 'shouldResetTransform')
        .mockImplementation(() => {
          return reset;
        });

      const context: StrictLayoutContext = {
        bubbles: true,
        strategy: manager.strategy,
        type: 'removed',
        target,
        targets,
        prevStrategy: undefined,
        stopPropagation() {
          this.bubbles = false;
        },
      };
      manager.onAfterLayout(context);

      expect(target.left).toBe(reset ? 0 : 50);
    });

    test('bubbling', () => {
      const targets = [new Group([new FabricObject()]), new FabricObject()];
      const target = new Group(targets);
      const parent = new Group([target]);
      const grandParent = new Group([parent]);
      const manager = new LayoutManager();

      const grandParentPerformLayout = jest.spyOn(
        grandParent.layoutManager,
        'performLayout'
      );

      const context: StrictLayoutContext = {
        bubbles: true,
        strategy: manager.strategy,
        type: 'added',
        target,
        targets,
        prevStrategy: undefined,
        stopPropagation() {
          this.bubbles = false;
        },
      };
      const layoutResult = {
        result: { centerX: 0, centerY: 0, width: 0, height: 0 },
        prevCenter: new Point(),
        nextCenter: new Point(),
        offset: new Point(),
      };
      manager.onAfterLayout(context, layoutResult);

      expect(grandParentPerformLayout.mock.calls[0]).toMatchObject([
        {
          type: 'added',
          targets,
          target: grandParent,
          path: [target, parent],
        },
      ]);
    });
  });
});
