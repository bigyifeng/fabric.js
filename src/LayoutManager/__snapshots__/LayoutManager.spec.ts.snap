// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout Manager getLayoutResult #type trigger 1`] = `
{
  "nextCenter": Point {
    "x": 50,
    "y": 100,
  },
  "offset": Point {
    "x": -70,
    "y": -120,
  },
  "prevCenter": Point {
    "x": 0,
    "y": 0,
  },
  "result": {
    "centerX": 50,
    "centerY": 100,
    "correctionX": 10,
    "correctionY": 20,
    "height": 250,
    "relativeCorrectionX": -30,
    "relativeCorrectionY": -40,
    "width": 200,
  },
}
`;

exports[`Layout Manager getLayoutResult #type trigger 2`] = `
{
  "nextCenter": Point {
    "x": 50,
    "y": 100,
  },
  "offset": Point {
    "x": 0,
    "y": 0,
  },
  "prevCenter": Point {
    "x": 0,
    "y": 0,
  },
  "result": {
    "centerX": 50,
    "centerY": 100,
    "correctionX": 10,
    "correctionY": 20,
    "height": 250,
    "relativeCorrectionX": -30,
    "relativeCorrectionY": -40,
    "width": 200,
  },
}
`;

exports[`Layout Manager getLayoutResult #type trigger 3`] = `
{
  "nextCenter": Point {
    "x": 50,
    "y": 100,
  },
  "offset": Point {
    "x": -70,
    "y": -120,
  },
  "prevCenter": Point {
    "x": 0,
    "y": 0,
  },
  "result": {
    "centerX": 50,
    "centerY": 100,
    "correctionX": 10,
    "correctionY": 20,
    "height": 250,
    "relativeCorrectionX": -30,
    "relativeCorrectionY": -40,
    "width": 200,
  },
}
`;
